File: Classes/CachedWebClient.vb
' Utility to automatically download radio programmes, using a plugin framework for provider specific implementation.
' Copyright © 2007-2010 Matt Robinson
'
' This program is free software; you can redistribute it and/or modify it under the terms of the GNU General
' Public License as published by the Free Software Foundation; either version 2 of the License, or (at your
' option) any later version.
'
' This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
' implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
' License for more details.
'
' You should have received a copy of the GNU General Public License along with this program; if not, write
' to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

Option Strict On
Option Explicit On

Imports System.Data.SQLite
Imports System.IO
Imports System.Net
Imports System.Runtime.Serialization.Formatters.Binary
Imports System.Text

Public Class CachedWebClient
    <Serializable()> _
    Public Structure CacheWebException
        Dim Message As String
        Dim InnerException As Exception
        Dim Status As WebExceptionStatus
        Dim Response As WebResponse
    End Structure

    Private Shared instance As New CachedWebClient

    <ThreadStatic()> _
    Private Shared dbConn As SQLiteConnection

    Public Shared Function GetInstance() As CachedWebClient
        Return instance
    End Function

    Private Function FetchDbConn() As SQLiteConnection
        If dbConn Is Nothing Then
            dbConn = New SQLiteConnection("Data Source=" + Path.Combine(FileUtils.GetAppDataFolder(), "httpcache.db") + ";Version=3")
            dbConn.Open()
        End If

        Return dbConn
    End Function

    Private Sub New()
        ' Create the table (and implicitly the database) for caching HTTP requests if it does not exist
        Using command As New SQLiteCommand("create table if not exists httpcache (uri varchar (1000) primary key, lastfetch datetime, success int, data blob)", FetchDbConn)
            command.ExecuteNonQuery()
        End Using

        ' Empty the table if it does exist
        Using command As New SQLiteCommand("delete from httpcache", FetchDbConn)
            command.ExecuteNonQuery()
        End Using
    End Sub

    Private Sub AddToHTTPCache(ByVal uri As String, ByVal requestSuccess As Boolean, ByVal data As Byte())
        Using command As New SQLiteCommand("insert or replace into httpcache (uri, lastfetch, success, data) values(@uri, @lastfetch, @success, @data)", FetchDbConn)
            command.Parameters.Add(New SQLiteParameter("@uri", uri))
            command.Parameters.Add(New SQLiteParameter("@lastfetch", Now))
            command.Parameters.Add(New SQLiteParameter("@success", requestSuccess))
            command.Parameters.Add(New SQLiteParameter("@data", data))
            command.ExecuteNonQuery()
        End Using
    End Sub

    Private Function GetHTTPCacheLastUpdate(ByVal uri As String) As Date
        Using command As New SQLiteCommand("select lastfetch from httpcache where uri=@uri", FetchDbConn)
            command.Parameters.Add(New SQLiteParameter("@uri", uri))

            Using reader As SQLiteDataReader = command.ExecuteReader
                If reader.Read Then
                    Return reader.GetDateTime(reader.GetOrdinal("lastfetch"))
                Else
                    Return Nothing
                End If
            End Using
        End Using
    End Function

    Private Function GetHTTPCacheContent(ByVal uri As String, ByRef requestSuccess As Boolean) As Byte()
        Using command As New SQLiteCommand("select success, data from httpcache where uri=@uri", FetchDbConn)
            command.Parameters.Add(New SQLiteParameter("@uri", uri))

            Using reader As SQLiteDataReader = command.ExecuteReader
                If reader.Read Then
                    requestSuccess = reader.GetBoolean(reader.GetOrdinal("success"))

                    ' Get the length of the content by passing nothing to getbytes
                    Dim contentLength As Integer = CInt(reader.GetBytes(reader.GetOrdinal("data"), 0, Nothing, 0, 0))

                    Dim content(contentLength - 1) As Byte
                    reader.GetBytes(reader.GetOrdinal("data"), 0, content, 0, contentLength)

                    Return content
                Else
                    Return Nothing
                End If
            End Using
        End Using
    End Function

    Public Function DownloadData(ByVal uri As String, ByVal fetchIntervalHrs As Integer) As Byte()
        If fetchIntervalHrs = 0 Then
            Throw New ArgumentException("fetchIntervalHrs cannot be zero.", "fetchIntervalHrs")
        End If

        Dim lastFetch As Date = GetHTTPCacheLastUpdate(uri)

        If lastFetch <> Nothing Then
            If lastFetch.AddHours(fetchIntervalHrs) > Now Then
                Dim requestSuccess As Boolean
                Dim cacheData As Byte() = GetHTTPCacheContent(uri, requestSuccess)

                If cacheData IsNot Nothing Then
                    If requestSuccess Then
                        Return cacheData
                    Else
                        Dim memoryStream As New MemoryStream(cacheData)
                        Dim binaryFormatter As New System.Runtime.Serialization.Formatters.Binary.BinaryFormatter()

                        ' Deserialise the CacheWebException structure
                        Dim cachedException As CacheWebException
                        cachedException = DirectCast(binaryFormatter.Deserialize(memoryStream), CacheWebException)

                        ' Crete a new WebException with the cached data and throw it
                        Throw New WebException(cachedException.Message, cachedException.InnerException, cachedException.Status, cachedException.Response)
                    End If
                End If
            End If
        End If

        Debug.Print("Cached WebClient: Fetching " + uri)
        Dim webClient As New WebClient
        webClient.Headers.Add("user-agent", My.Application.Info.AssemblyName + " " + My.Application.Info.Version.ToString)

        Dim data As Byte()

        Try
            data = webClient.DownloadData(uri)
        Catch webExp As WebException
            If webExp.Status <> WebExceptionStatus.NameResolutionFailure And webExp.Status <> WebExceptionStatus.Timeout Then
                ' A WebException doesn't serialise well, as Response and Status get lost,
                ' so store the information in a structure and then recreate it later
                Dim cacheException As New CacheWebException
                cacheException.Message = webExp.Message
                cacheException.InnerException = webExp.InnerException
                cacheException.Status = webExp.Status
                cacheException.Response = webExp.Response

                Dim stream As New MemoryStream()
                Dim formatter As New BinaryFormatter()

                ' Serialise the CacheWebException and store it in the cache
                formatter.Serialize(stream, cacheException)
                AddToHTTPCache(uri, False, stream.ToArray())
            End If

            ' Re-throw the WebException
            Throw
        End Try

        AddToHTTPCache(uri, True, data)

        Return data
    End Function

    Public Function DownloadString(ByVal uri As String, ByVal fetchIntervalHrs As Integer) As String
        Return Encoding.UTF8.GetString(DownloadData(uri, fetchIntervalHrs))
    End Function
End Class

File: Classes/CachedWebClient.cs
/*
 * Copyright © 2008-2018 Matt Robinson
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

namespace RadioDld
{
    using System;
    using System.Data.SQLite;
    using System.Diagnostics;
    using System.IO;
    using System.IO.Compression;
    using System.Net;
    using System.Runtime.Serialization.Formatters.Binary;

    internal class CachedWebClient : CachedWebClientBase
    {
        [ThreadStatic]
        private static SQLiteConnection dbConn;

        static CachedWebClient()
        {
            File.Delete(DatabasePath());

            // Create the the database and table for caching HTTP requests
            using (SQLiteCommand command = new SQLiteCommand("create table httpcache (uri varchar (1000) primary key, lastfetch datetime, success int, data blob)", FetchDbConn()))
            {
                command.ExecuteNonQuery();
            }
        }

        public override byte[] DownloadData(Uri uri, int fetchIntervalHrs, string userAgent)
        {
            if (fetchIntervalHrs == 0)
            {
                throw new ArgumentException("fetchIntervalHrs cannot be zero.", "fetchIntervalHrs");
            }

            DateTime? lastFetch = this.GetHTTPCacheLastUpdate(uri);

            if (lastFetch != null)
            {
                if (lastFetch.Value.AddHours(fetchIntervalHrs) > DateTime.Now)
                {
                    bool requestSuccess = false;
                    byte[] cacheData = this.GetHTTPCacheContent(uri, ref requestSuccess);

                    if (cacheData != null)
                    {
                        if (requestSuccess)
                        {
                            return cacheData;
                        }
                        else
                        {
                            using (MemoryStream memoryStream = new MemoryStream(cacheData))
                            {
                                BinaryFormatter binaryFormatter = new BinaryFormatter();

                                // Deserialise the CacheWebException structure
                                CacheWebExpInfo cachedException = default(CacheWebExpInfo);
                                cachedException = (CacheWebExpInfo)binaryFormatter.Deserialize(memoryStream);

                                // Crete a new WebException with the cached data and throw it
                                throw new WebException(cachedException.Message, cachedException.InnerException, cachedException.Status, cachedException.Response);
                            }
                        }
                    }
                }
            }

            Debug.Print("CachedWebClient: Fetching " + uri.ToString());

            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
            request.UserAgent = userAgent;
            request.Headers.Add(HttpRequestHeader.AcceptEncoding, "gzip");

            using (MemoryStream dataStream = new MemoryStream())
            {
                try
                {
                    HttpWebResponse response = (HttpWebResponse)request.GetResponse();
                    Stream responseStream = response.GetResponseStream();

                    if (response.ContentEncoding.ToUpperInvariant() == "GZIP")
                    {
                        // Decompress the gzipped response while it is being read into the memory stream
                        responseStream = new GZipStream(responseStream, CompressionMode.Decompress);
                    }

                    // Read the response into a MemoryStream in chunks, ready for the byte array
                    byte[] buffer = new byte[10240];
                    int readBytes;

                    while ((readBytes = responseStream.Read(buffer, 0, buffer.Length)) > 0)
                    {
                        dataStream.Write(buffer, 0, readBytes);
                    }

                    if (response.ContentEncoding.ToUpperInvariant() == "GZIP")
                    {
                        responseStream.Dispose();
                    }
                }
                catch (WebException webExp)
                {
                    if (webExp.Status != WebExceptionStatus.NameResolutionFailure && webExp.Status != WebExceptionStatus.Timeout)
                    {
                        // A WebException doesn't serialise well, as Response and Status get lost,
                        // so store the information in a structure and then recreate it later
                        CacheWebExpInfo cacheException = new CacheWebExpInfo();
                        cacheException.Message = webExp.Message;
                        cacheException.InnerException = webExp.InnerException;
                        cacheException.Status = webExp.Status;
                        cacheException.Response = webExp.Response;

                        using (MemoryStream stream = new MemoryStream())
                        {
                            BinaryFormatter formatter = new BinaryFormatter();

                            // Serialise the CacheWebException and store it in the cache
                            formatter.Serialize(stream, cacheException);
                            this.AddToHTTPCache(uri, false, stream.ToArray());
                        }
                    }

                    // Re-throw the WebException
                    throw;
                }

                byte[] data = dataStream.ToArray();
                this.AddToHTTPCache(uri, true, data);

                return data;
            }
        }

        private static string DatabasePath()
        {
            return Path.Combine(Path.GetTempPath(), "radiodld-httpcache.db");
        }

        private static SQLiteConnection FetchDbConn()
        {
            if (dbConn == null)
            {
                dbConn = new SQLiteConnection("Data Source=" + DatabasePath() + ";Version=3");
                dbConn.Open();
            }

            return dbConn;
        }

        private void AddToHTTPCache(Uri uri, bool requestSuccess, byte[] data)
        {
            using (SQLiteCommand command = new SQLiteCommand("insert or replace into httpcache (uri, lastfetch, success, data) values(@uri, @lastfetch, @success, @data)", FetchDbConn()))
            {
                command.Parameters.Add(new SQLiteParameter("@uri", uri.ToString()));
                command.Parameters.Add(new SQLiteParameter("@lastfetch", DateTime.Now));
                command.Parameters.Add(new SQLiteParameter("@success", requestSuccess));
                command.Parameters.Add(new SQLiteParameter("@data", data));
                command.ExecuteNonQuery();
            }
        }

        private DateTime? GetHTTPCacheLastUpdate(Uri uri)
        {
            using (SQLiteCommand command = new SQLiteCommand("select lastfetch from httpcache where uri=@uri", FetchDbConn()))
            {
                command.Parameters.Add(new SQLiteParameter("@uri", uri.ToString()));

                using (SQLiteDataReader reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        return reader.GetDateTime(reader.GetOrdinal("lastfetch"));
                    }
                    else
                    {
                        return null;
                    }
                }
            }
        }

        private byte[] GetHTTPCacheContent(Uri uri, ref bool requestSuccess)
        {
            using (SQLiteCommand command = new SQLiteCommand("select success, data from httpcache where uri=@uri", FetchDbConn()))
            {
                command.Parameters.Add(new SQLiteParameter("@uri", uri.ToString()));

                using (SQLiteDataReader reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        requestSuccess = reader.GetBoolean(reader.GetOrdinal("success"));

                        // Get the length of the content by passing nothing to getbytes
                        int contentLength = Convert.ToInt32(reader.GetBytes(reader.GetOrdinal("data"), 0, null, 0, 0));

                        byte[] content = new byte[contentLength];
                        reader.GetBytes(reader.GetOrdinal("data"), 0, content, 0, contentLength);

                        return content;
                    }
                    else
                    {
                        return null;
                    }
                }
            }
        }

        [Serializable]
        public class CacheWebExpInfo
        {
            public string Message { get; set; }

            public Exception InnerException { get; set; }

            public WebExceptionStatus Status { get; set; }

            public WebResponse Response { get; set; }
        }
    }
}
